src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -0,0 +1,411 @@
+        var course = courseRepository.findWithEagerLearningPathsAndCompetenciesByIdElseThrow(courseId);
+        var learningPaths = course.getLearningPaths();
+        learningPaths.forEach(learningPath -> learningPath.addCompetency(competency));
+        learningPathRepository.saveAll(learningPaths);
+        log.debug("Linked competency (id={}) to learning paths", competency.getId());
+    }
+
+    /**
+     * Remove linked competency from all learning paths of the course.
+     *
+     * @param competency Competency that should be removed from each learning path
+     * @param courseId   course id that the learning paths belong to
+     */
+    public void removeLinkedCompetencyFromLearningPathsOfCourse(Competency competency, long courseId) {
+        var course = courseRepository.findWithEagerLearningPathsAndCompetenciesByIdElseThrow(courseId);
+        var learningPaths = course.getLearningPaths();
+        learningPaths.forEach(learningPath -> learningPath.removeCompetency(competency));
+        learningPathRepository.saveAll(learningPaths);
+        log.debug("Removed linked competency (id={}) from learning paths", competency.getId());
+    }
+
+    public void updateLearningPathProgress(final long learningPathId) {
+        final var learningPath = learningPathRepository.findWithEagerCompetenciesByIdElseThrow(learningPathId);
+        this.updateLearningPathProgress(learningPath);
+    }
+
+    /**
+     * Updates progress of the learning path specified by course and user id.
+     *
+     * @param courseId id of the course the learning path is linked to
+     * @param userId   id of the user the learning path is linked to
+     */
+    public void updateLearningPathProgress(final long courseId, final long userId) {
+        final var learningPath = learningPathRepository.findWithEagerCompetenciesByCourseIdAndUserId(courseId, userId);
+        learningPath.ifPresent(this::updateLearningPathProgress);
+    }
+
+    /**
+     * Updates progress of the given learning path. Competencies of the learning path must be loaded eagerly.
+     *
+     * @param learningPath learning path that is updated
+     */
+    private void updateLearningPathProgress(final LearningPath learningPath) {
+        final var userId = learningPath.getUser().getId();
+        final var competencyIds = learningPath.getCompetencies().stream().map(Competency::getId).collect(Collectors.toSet());
+        final var competencyProgresses = competencyProgressRepository.findAllByCompetencyIdsAndUserId(competencyIds, userId);
+
+        final var completed = (float) competencyProgresses.stream().filter(CompetencyProgressService::isMastered).count();
+        final var numberOfCompetencies = learningPath.getCompetencies().size();
+        if (numberOfCompetencies == 0) {
+            learningPath.setProgress(0);
+        }
+        else {
+            learningPath.setProgress(Math.round(completed * 100 / (float) numberOfCompetencies));
```

Comment:
One of the float casts is irrelevant.
And with actual types instead of `var` you could also get rid of the first cast.

Any reason why you don't use double?

Reply:
Why floats: `Math.round(double)` returns a  `long`. This would require a downcast or the type of the DB should be changed. Since the values are 0-100 and the precision is fairly irrelevant using `Math.round(float) -> int` seems fitting in that case.

Removed the irrelevant casts.